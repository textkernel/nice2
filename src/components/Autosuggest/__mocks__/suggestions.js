export const SUGGESTIONS = [
    { name: 'Sun' },
    { name: 'Moon' },
    { name: 'Mercury' },
    { name: 'Venus' },
    { name: 'Earth' },
    { name: 'Mars' },
    { name: 'Jupiter' },
    { name: 'Saturn' },
    { name: 'Neptun' }
];

export const SUGGESTION_TO_STRING = item => (item ? item.name : '');
