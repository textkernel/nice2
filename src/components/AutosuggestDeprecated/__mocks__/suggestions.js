export const SUGGESTIONS = [
    { name: 'Sun' },
    { name: 'Moon' },
    { name: 'Mercury' },
    { name: 'Venus' },
    { name: 'Earth' },
    { name: 'Mars' },
    { name: 'Jupiter' },
    { name: 'Saturn' },
    { name: 'Neptun' },
];

export const COMPLEX_SUGGESTIONS = [
    { name: 'Sun', type: 'Star', count: 1234 },
    { name: 'Moon', type: 'Star', count: 1234 },
    { name: 'Mercury', type: 'Planet', count: 1234 },
    { name: 'Venus', type: 'Planet', count: 1234 },
    { name: 'Earth', type: 'Planet', count: 1234 },
    { name: 'Mars', type: 'Planet', count: 1234 },
    { name: 'Jupiter', type: 'Planet', count: 1234 },
    { name: 'Saturn', type: 'Planet', count: 1234 },
    { name: 'Neptun', type: 'Planet', count: 1234 },
];

export const SUGGESTION_TO_STRING = (item) => (item ? item.name : '');
