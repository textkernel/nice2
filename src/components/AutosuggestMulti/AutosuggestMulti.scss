.AutosuggestMulti {
    &__wrapper {
        cursor: pointer;
        display: flex;
        flex-wrap: wrap;
        flex-direction: row;
        align-items: stretch;
        overflow: hidden;
        min-height: 30px;
        outline: none;
        padding: var(--spacing-normal);

        &--isFocused {
            height: auto;
            flex-wrap: wrap;
        }
    }

    &__input {
        border: 0;
        margin: var(--spacing-normal);
        flex: 1;
        font-size: var(--font-size-normal);
        min-width: 100px;

        &:focus {
            outline: none;
        }

        &::placeholder {
            color: var(--color-muted);
        }
    }
}
