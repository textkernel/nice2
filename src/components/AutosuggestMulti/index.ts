export { AutosuggestMulti } from './AutosuggestMulti';
export { SuggestionTag } from './SuggestionTag';
