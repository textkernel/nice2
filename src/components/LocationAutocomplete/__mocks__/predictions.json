[
    {
        "description": "Truro, UK",
        "id": "51b6747d060368917397ce3cc25abe2089302262",
        "matched_substrings": [
            {
                "length": 2,
                "offset": 0
            }
        ],
        "place_id": "ChIJdRpa1XwQa0gRtAcdle9HY2E",
        "reference": "ChIJdRpa1XwQa0gRtAcdle9HY2E",
        "structured_formatting": {
            "main_text": "Truro",
            "main_text_matched_substrings": [
                {
                    "length": 2,
                    "offset": 0
                }
            ],
            "secondary_text": "UK"
        },
        "terms": [
            {
                "offset": 0,
                "value": "Truro"
            },
            {
                "offset": 7,
                "value": "UK"
            }
        ],
        "types": ["locality", "political", "geocode"]
    },
    {
        "description": "Trowbridge, UK",
        "id": "b10d478a406230745e98edd2d2a01f59097d5a59",
        "matched_substrings": [
            {
                "length": 2,
                "offset": 0
            }
        ],
        "place_id": "ChIJM6W0Ewp9cUgRLXIknC4a-Pw",
        "reference": "ChIJM6W0Ewp9cUgRLXIknC4a-Pw",
        "structured_formatting": {
            "main_text": "Trowbridge",
            "main_text_matched_substrings": [
                {
                    "length": 2,
                    "offset": 0
                }
            ],
            "secondary_text": "UK"
        },
        "terms": [
            {
                "offset": 0,
                "value": "Trowbridge"
            },
            {
                "offset": 12,
                "value": "UK"
            }
        ],
        "types": ["locality", "political", "geocode"]
    },
    {
        "description": "Tring, UK",
        "id": "8b1f6899d3555c6aae3a2d383ed96c7649f3ccf8",
        "matched_substrings": [
            {
                "length": 2,
                "offset": 0
            }
        ],
        "place_id": "ChIJ33SmFitYdkgR3-ErzH_p6Bw",
        "reference": "ChIJ33SmFitYdkgR3-ErzH_p6Bw",
        "structured_formatting": {
            "main_text": "Tring",
            "main_text_matched_substrings": [
                {
                    "length": 2,
                    "offset": 0
                }
            ],
            "secondary_text": "UK"
        },
        "terms": [
            {
                "offset": 0,
                "value": "Tring"
            },
            {
                "offset": 7,
                "value": "UK"
            }
        ],
        "types": ["locality", "political", "geocode"]
    },
    {
        "description": "Troon, UK",
        "id": "c563f4c69998a376dac8e3b4122067d782e752f7",
        "matched_substrings": [
            {
                "length": 2,
                "offset": 0
            }
        ],
        "place_id": "ChIJxy8qU3jRiUgR-sKnFLHxiSM",
        "reference": "ChIJxy8qU3jRiUgR-sKnFLHxiSM",
        "structured_formatting": {
            "main_text": "Troon",
            "main_text_matched_substrings": [
                {
                    "length": 2,
                    "offset": 0
                }
            ],
            "secondary_text": "UK"
        },
        "terms": [
            {
                "offset": 0,
                "value": "Troon"
            },
            {
                "offset": 7,
                "value": "UK"
            }
        ],
        "types": ["locality", "political", "geocode"]
    },
    {
        "description": "Tredegar, UK",
        "id": "4dca94194f54dee0879b9d368379eb5d09ec838d",
        "matched_substrings": [
            {
                "length": 2,
                "offset": 0
            }
        ],
        "place_id": "ChIJdQjXYyYkbkgRTRfZ8WAt6lI",
        "reference": "ChIJdQjXYyYkbkgRTRfZ8WAt6lI",
        "structured_formatting": {
            "main_text": "Tredegar",
            "main_text_matched_substrings": [
                {
                    "length": 2,
                    "offset": 0
                }
            ],
            "secondary_text": "UK"
        },
        "terms": [
            {
                "offset": 0,
                "value": "Tredegar"
            },
            {
                "offset": 10,
                "value": "UK"
            }
        ],
        "types": ["locality", "political", "geocode"]
    }
]
