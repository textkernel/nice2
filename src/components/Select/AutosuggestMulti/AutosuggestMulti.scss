.AutosuggestMulti {
    &__wrapper {
        cursor: pointer;
        display: flex;
        flex-wrap: nowrap;
        flex-direction: row;
        align-items: stretch;
        overflow: hidden;
        height: 34px;
        box-sizing: border-box;
        outline: none;
        padding: var(--spacing-normal);
        padding-bottom: var(--margin);

        &--isFocused {
            min-height: 34px;
            height: auto;
            flex-wrap: wrap;
        }
    }

    &__input {
        border: 0;
        margin: var(--spacing-normal);
        padding-left: var(--spacing-normal);
        flex: 1;
        font-size: var(--font-size-normal);
        min-width: 105px;

        &--hidden {
            position: absolute;
            opacity: 0;
        }

        &:focus {
            outline: none;
        }

        &::placeholder {
            color: var(--color-muted);
        }
    }
}
