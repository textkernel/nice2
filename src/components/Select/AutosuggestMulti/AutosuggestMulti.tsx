import * as React from 'react';
import { bem } from '../../../utils';
import { SuggestionTag } from './SuggestionTag';
import { SuggestionsList } from '../SuggestionsList';
import {
    SelectBase,
    CommonPropsWithClear,
    FocusedRendererHelpers,
    BlurredRendererHelpers,
} from '../SelectBase';
import styles from './AutosuggestMulti.scss';
import { BACKSPACE_KEY, ESCAPE_KEY, TAB_KEY, ENTER_KEY } from '../../../constants';

interface Props<S> extends CommonPropsWithClear<S> {
    /** define id for input element */
    id?: string;
    /** array of already selected suggestions */
    selectedSuggestions: S[];
    /** number of visible tags in blur mode */
    numberOfVisibleTags: number;
    /** to be shown in the input field when no value is typed */
    inputPlaceholder: string;
    /** Enable ListOptimizer component for decreasing render time */
    useOptimizeListRender?: boolean;
    /** onSelectionChange() called when a suggestion is removed  */
    onSelectionRemove: (item: S) => void;
}

const { elem } = bem('AutosuggestMulti', styles);

export function AutosuggestMulti<S>(props: Props<S>) {
    const {
        id,
        onInputValueChange,
        onSelectionChange,
        selectedSuggestions,
        suggestionToString,
        inputPlaceholder,
        useOptimizeListRender,
        suggestions,
        numberOfVisibleTags,
        onBlur,
        showClearButton,
        onSelectionRemove,
        ...rest
    } = props;
    const inputRef = React.createRef<HTMLInputElement>();
    const [inputValue, setInputValue] = React.useState('');
    const [selectOnBlur, setSelectOnBlur] = React.useState(false);

    const handleInputValueChange = (value: string) => {
        onInputValueChange(value);
        setInputValue(value);
    };

    const renderFullTagsList = () => {
        return selectedSuggestions.map((item) => (
            <SuggestionTag key={suggestionToString(item)} onClick={() => onSelectionRemove(item)}>
                {suggestionToString(item)}
            </SuggestionTag>
        ));
    };

    const renderShortTagsList = () => {
        const visibleTags = selectedSuggestions.slice(0, numberOfVisibleTags);
        const hiddenTags = selectedSuggestions.slice(numberOfVisibleTags);
        const numberOfHiddenTags = hiddenTags.length;
        const visibleTagsList = visibleTags.map((item) => (
            <SuggestionTag
                key={suggestionToString(item)}
                width={numberOfHiddenTags > 0 ? 'block' : 'auto'}
            >
                {suggestionToString(item)}
            </SuggestionTag>
        ));

        if (numberOfHiddenTags > 0) {
            const counter = (
                <SuggestionTag key="counter" width="small">
                    {`+${numberOfHiddenTags}`}
                </SuggestionTag>
            );
            return [...visibleTagsList, counter];
        }

        return visibleTagsList;
    };

    const handleInputKeyDown = (blur: () => void) => (event: React.KeyboardEvent<HTMLElement>) => {
        if (event.key === TAB_KEY) {
            setSelectOnBlur(true);
            blur();
        } else if (event.key === ENTER_KEY && !inputValue) {
            /**
             * Prevent the default Downshift handler behavior
             * That need for submitting form
             */
            // eslint-disable-next-line no-param-reassign, dot-notation
            event.nativeEvent['preventDownshiftDefault'] = true;
        } else if (event.key === ESCAPE_KEY) {
            setSelectOnBlur(false);
            inputRef.current?.blur();
            blur();
        } else if (event.key === BACKSPACE_KEY && !inputValue && !!selectedSuggestions.length) {
            const lastItem = selectedSuggestions[selectedSuggestions.length - 1];
            onSelectionRemove(lastItem);
        }
    };

    // eslint-disable-next-line react/display-name
    const renderFocused: FocusedRendererHelpers<S> = ({ getInputProps, onBlur: blur }) => (
        <div {...elem('wrapper', { isFocused: true })}>
            {renderFullTagsList()}
            <input
                {...getInputProps({
                    id,
                    ref: inputRef,
                    placeholder: inputPlaceholder,
                    onKeyDown: handleInputKeyDown(blur),
                    'data-lpignore': true,
                    ...elem('input'),
                })}
            />
        </div>
    );

    // eslint-disable-next-line react/display-name
    const renderBlurred: BlurredRendererHelpers<S> = ({ getInputProps, onFocus }) => (
        <div {...elem('wrapper')}>
            {renderShortTagsList()}
            <input
                {...getInputProps({
                    id,
                    ref: inputRef,
                    placeholder: selectedSuggestions.length === 0 ? inputPlaceholder : '',
                    'data-lpignore': true,
                    onFocus,
                    ...elem('input', { hidden: selectedSuggestions.length > 0 }),
                })}
            />
        </div>
    );

    return (
        <SelectBase
            {...rest}
            suggestions={suggestions}
            suggestionToString={suggestionToString}
            inputRef={inputRef}
            onBlur={onBlur}
            onSelectionChange={onSelectionChange}
            onInputValueChange={handleInputValueChange}
            listRenderer={(listProps) => (
                <SuggestionsList {...listProps} useOptimizeRender={useOptimizeListRender} />
            )}
            focusedRenderer={renderFocused}
            blurredRenderer={renderBlurred}
            showClearButton={showClearButton && selectedSuggestions.length > 0}
            keepExpandedAfterSelection
            selectOnBlur={selectOnBlur}
            clearInputAfterSelection
        />
    );
}

AutosuggestMulti.displayName = 'AutosuggestMulti';

AutosuggestMulti.defaultProps = {
    numberOfVisibleTags: 3,
    selectedSuggestions: [],
};
