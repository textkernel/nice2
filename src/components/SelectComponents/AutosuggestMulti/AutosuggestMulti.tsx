import * as React from 'react';
import { bem } from '../../../utils';
import { SuggestionTag } from './SuggestionTag';
import { SuggestionsList } from '../SuggestionsList';
import {
    SelectBase,
    BasicSelectProps,
    SelectClearButtonProps,
    SelectInputFieldProps,
    FocusedRendererHelpers,
    BlurredRendererHelpers,
    ListRendererHelper,
} from '../SelectBase';
import styles from './AutosuggestMulti.scss';
import { BACKSPACE_KEY, ESCAPE_KEY, ENTER_KEY } from '../../../constants';

interface Props<S>
    extends BasicSelectProps<S>,
        SelectClearButtonProps,
        Omit<SelectInputFieldProps, 'clearInputAfterSelection'> {
    /** HTML id for the input element */
    id?: string;
    /** Creates a unique (React) key for a suggestion item. If undefined suggestionToString will be used */
    suggestionToKey?: (suggestions: S) => string;
    /** An array of already selected suggestions */
    selectedSuggestions?: S[];
    /** Number of visible tags in blur mode */
    numberOfVisibleTags?: number;
    /** String to be shown in the input field when no value is typed */
    inputPlaceholder: string;
    /** String to be shown when no suggestions are available */
    noSuggestionsPlaceholder?: string;
    /** Defines if the first item of suggestions list is visible even while loading other elements */
    isFirstItemAlwaysVisible?: boolean;
    /** Enable ListOptimizer component for decreasing render time */
    useOptimizeListRender?: boolean;
    /** Function to be called when a suggestion is removed  */
    onSelectionRemove?: (item: S) => void;
    /** Function to be called on submitting form */
    onSubmit?: () => void;
    /** if suggestions are still loading, i.e. display placeholders */
    isLoading?: boolean;
}

const { elem } = bem('AutosuggestMulti', styles);

export function AutosuggestMulti<S>(props: Props<S>) {
    const {
        id,
        onInputValueChange,
        onSelectionAdd,
        selectedSuggestions = [],
        suggestionToString,
        suggestionToKey,
        suggestionItemRenderer,
        inputPlaceholder,
        noSuggestionsPlaceholder,
        useOptimizeListRender,
        suggestions,
        isLoading,
        numberOfVisibleTags,
        isFirstItemAlwaysVisible,
        onFocus,
        onBlur,
        onSubmit,
        disabled,
        showClearButton,
        onSelectionRemove,
        ...rest
    } = props;
    const inputRef = React.createRef<HTMLInputElement>();
    const [inputValue, setInputValue] = React.useState('');

    const handleInputValueChange = (value: string) => {
        onInputValueChange?.(value);
        setInputValue(value);
    };

    const renderFullTagsList = () => {
        return selectedSuggestions.map((item) => (
            <SuggestionTag key={suggestionToString(item)} onClick={() => onSelectionRemove?.(item)}>
                {suggestionToString(item)}
            </SuggestionTag>
        ));
    };

    const renderShortTagsList = () => {
        const visibleTags = selectedSuggestions.slice(0, numberOfVisibleTags);
        const hiddenTags = selectedSuggestions.slice(numberOfVisibleTags);
        const numberOfHiddenTags = hiddenTags.length;
        const visibleTagsList = visibleTags.map((item) => (
            <SuggestionTag
                key={suggestionToString(item)}
                width={numberOfHiddenTags > 0 ? 'block' : 'auto'}
            >
                {suggestionToString(item)}
            </SuggestionTag>
        ));

        if (numberOfHiddenTags > 0) {
            const counter = (
                <SuggestionTag key="counter" width="small">
                    {`+${numberOfHiddenTags}`}
                </SuggestionTag>
            );
            return [...visibleTagsList, counter];
        }

        return visibleTagsList;
    };

    const handleInputKeyDown = (
        event: React.KeyboardEvent<HTMLElement>,
        highlightedIndex: number | null
    ) => {
        const isHighlighted = highlightedIndex !== null && highlightedIndex > -1;
        if (event.key === ENTER_KEY && !inputValue && !isHighlighted) {
            /**
             * Prevent the default Downshift handler behavior
             * That need for submitting form
             */
            // eslint-disable-next-line no-param-reassign, dot-notation
            event.nativeEvent['preventDownshiftDefault'] = true;
            onSubmit?.();
        } else if (event.key === ESCAPE_KEY) {
            inputRef.current?.blur();
        } else if (event.key === BACKSPACE_KEY && !inputValue && !!selectedSuggestions.length) {
            const lastItem = selectedSuggestions[selectedSuggestions.length - 1];
            onSelectionRemove?.(lastItem);
        }
    };

    // eslint-disable-next-line react/display-name
    const renderFocused: FocusedRendererHelpers<S> = ({
        getInputProps,
        onFocus: onFocusInput,
        highlightedIndex,
    }) => (
        <div {...elem('wrapper', { disabled, isFocused: true })}>
            {renderFullTagsList()}
            <input
                {...getInputProps({
                    id,
                    disabled,
                    ref: inputRef,
                    placeholder: inputPlaceholder,
                    onFocus: onFocusInput,
                    onKeyDown: (e) => handleInputKeyDown(e, highlightedIndex),
                    'data-lpignore': true,
                    ...elem('input', { disabled }),
                })}
            />
        </div>
    );

    // eslint-disable-next-line react/display-name
    const renderBlurred: BlurredRendererHelpers<S> = ({ getInputProps, onFocus: onFocusInput }) => (
        <div {...elem('wrapper', { disabled })}>
            {renderShortTagsList()}
            <input
                {...getInputProps({
                    id,
                    disabled,
                    ref: inputRef,
                    placeholder: selectedSuggestions.length === 0 ? inputPlaceholder : '',
                    'data-lpignore': true,
                    onFocus: onFocusInput,
                    ...elem('input', { disabled, hidden: selectedSuggestions.length > 0 }),
                })}
            />
        </div>
    );

    const renderList: ListRendererHelper<S> = (listProps) => {
        const { inputValue: inputToList } = listProps;

        return inputToList ? (
            <SuggestionsList
                {...listProps}
                isFirstItemAlwaysVisible={isFirstItemAlwaysVisible}
                isLoading={isLoading}
                useOptimizeRender={useOptimizeListRender}
                suggestionToKey={suggestionToKey}
                suggestionItemRenderer={suggestionItemRenderer}
                noSuggestionsPlaceholder={noSuggestionsPlaceholder}
            />
        ) : null;
    };

    return (
        <SelectBase
            {...rest}
            suggestions={suggestions}
            suggestionToString={suggestionToString}
            inputRef={inputRef}
            disabled={disabled}
            onFocus={onFocus}
            onBlur={onBlur}
            onSelectionAdd={onSelectionAdd}
            onInputValueChange={handleInputValueChange}
            listRenderer={renderList}
            focusedRenderer={renderFocused}
            blurredRenderer={renderBlurred}
            showClearButton={showClearButton && selectedSuggestions.length > 0}
            highlightOnEmptyInput={false}
            keepExpandedAfterSelection
            selectOnTab
            clearInputAfterSelection
        />
    );
}

AutosuggestMulti.displayName = 'AutosuggestMulti';

AutosuggestMulti.defaultProps = {
    id: undefined,
    numberOfVisibleTags: 3,
    selectedSuggestions: [],
    suggestionToKey: null,
    isFirstItemAlwaysVisible: false,
    useOptimizeListRender: false,
    onSubmit: null,
    onSelectionRemove: null,
    noSuggestionsPlaceholder: '',
    isLoading: false,
};
