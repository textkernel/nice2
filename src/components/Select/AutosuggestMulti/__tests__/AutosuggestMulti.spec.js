import React from 'react';
import toJson from 'enzyme-to-json';
import { AutosuggestMulti } from '../AutosuggestMulti';
import { SUGGESTIONS, SUGGESTION_TO_STRING } from '../../../Autosuggest/__mocks__/suggestions';

describe('AutosuggestMulti', () => {
    const suggestionToString = SUGGESTION_TO_STRING;
    const inputPlaceholder = 'type here...';
    const numberOfVisibleTags = 3;
    const mockOnSelectionChange = jest.fn();
    const mockOnSelectionRemove = jest.fn();
    const mockOnInputValueChange = jest.fn();
    const mockOnBlur = jest.fn();

    let suggestionsList = [];
    let selectedSuggestions = [];

    let wrapper;
    const setFocusOnInput = () => wrapper.find('input').simulate('click');

    beforeEach(() => {
        wrapper = mount(
            <AutosuggestMulti
                isLoading={false}
                selectedSuggestions={selectedSuggestions}
                suggestions={suggestionsList}
                suggestionToString={suggestionToString}
                inputPlaceholder={inputPlaceholder}
                onSelectionChange={mockOnSelectionChange}
                onSelectionRemove={mockOnSelectionRemove}
                onInputValueChange={mockOnInputValueChange}
                numberOfVisibleTags={numberOfVisibleTags}
                onBlur={mockOnBlur}
            />
        );
    });

    afterEach(() => {
        suggestionsList = [];
        selectedSuggestions = [];
        wrapper.unmount();
    });

    describe('rendering', () => {
        it('should initially render empty component correctly', () => {
            expect(toJson(wrapper)).toMatchSnapshot();
        });
        it('should render component with suggestions', () => {
            suggestionsList = SUGGESTIONS.slice(1, 20);
            wrapper.setProps({ suggestions: suggestionsList });
            setFocusOnInput();
            wrapper.find('input').simulate('change', { target: { value: 'driver' } });
            expect(toJson(wrapper)).toMatchSnapshot();
        });
        it('should render isLoading state', () => {
            suggestionsList = SUGGESTIONS.slice(1, 20);
            wrapper.setProps({ suggestions: [], isLoading: true });
            setFocusOnInput();
            wrapper.find('input').simulate('change', { target: { value: 'driver' } });
            expect(wrapper.find('.SuggestionsList__loaderItem')).toHaveLength(5);
            expect(toJson(wrapper)).toMatchSnapshot();
        });
        it('should render empty component correctly when focused', async () => {
            setFocusOnInput();
            const inputNode = wrapper.find('input').getDOMNode();
            expect(document.activeElement).toBe(inputNode);
        });
        it('should render tag for each selected selection when component is focused', () => {
            selectedSuggestions = SUGGESTIONS.slice(0, 5);
            wrapper.setProps({ selectedSuggestions });
            setFocusOnInput();

            expect(wrapper.find('SuggestionTag')).toHaveLength(selectedSuggestions.length);
        });
        it('should render tag for each selected selection when component is blurred', () => {
            selectedSuggestions = SUGGESTIONS.slice(0, 5);
            wrapper.setProps({ selectedSuggestions });

            expect(wrapper.find('SuggestionTag')).toHaveLength(numberOfVisibleTags + 1);
        });
    });
    describe('focusing and blurring the search field', () => {
        it('should clear input value on pressing Escape button', (done) => {
            const textInputValue = 'driver';
            setFocusOnInput();
            wrapper.find('input').simulate('change', { target: { value: textInputValue } });
            expect(wrapper.find('input').props().value).toEqual(textInputValue);
            wrapper.find('input').simulate('blur').simulate('keyDown', { key: 'Escape' });
            setTimeout(() => {
                wrapper.update();
                expect(wrapper.find('input').props().value).toEqual('');
                done();
            });
        });
        it('should blur on pressing Tab button', (done) => {
            setFocusOnInput();
            wrapper.find('input').simulate('blur').simulate('keyDown', { key: 'Tab' });
            setTimeout(() => {
                wrapper.update();
                expect(mockOnBlur).toHaveBeenCalled();
                done();
            });
        });
    });
    describe('callbacks', () => {
        describe('onSelectionChange', () => {
            it('should be called on clicking on a suggestion', () => {
                suggestionsList = SUGGESTIONS.slice(1, 20);
                wrapper.setProps({ suggestions: suggestionsList });
                setFocusOnInput();

                expect(mockOnSelectionChange).not.toHaveBeenCalled();

                wrapper.find('input').simulate('change', { target: { value: 'a' } });
                wrapper.find('li').first().children().simulate('click');

                expect(mockOnSelectionChange).toHaveBeenCalled();
            });
            it('should be called also when clicking on a suggestion the second time in a row', () => {
                suggestionsList = SUGGESTIONS.slice(1, 20);
                wrapper.setProps({ suggestions: suggestionsList });
                setFocusOnInput();

                expect(mockOnSelectionChange).not.toHaveBeenCalled();

                wrapper.find('li').first().children().simulate('click');

                expect(mockOnSelectionChange).toHaveBeenCalledTimes(1);

                setFocusOnInput();
                wrapper.find('li').first().children().simulate('click');

                expect(mockOnSelectionChange).toHaveBeenCalledTimes(2);
            });
            it('should be called on deleting a suggestion by clicking on the x button next to it', () => {
                selectedSuggestions = SUGGESTIONS.slice(0, 5);
                wrapper.setProps({ selectedSuggestions });
                setFocusOnInput();

                expect(mockOnSelectionChange).not.toHaveBeenCalled();
                expect(wrapper.find('SuggestionTag')).toHaveLength(selectedSuggestions.length);

                wrapper.find('SuggestionTag').at(2).find('button').simulate('click');

                expect(mockOnSelectionRemove).toHaveBeenCalledWith(selectedSuggestions[2]);
            });
            it('should be called on deleting a suggestion by hitting backspace in the empty input field', () => {
                selectedSuggestions = SUGGESTIONS.slice(0, 5);
                wrapper.setProps({ selectedSuggestions });
                setFocusOnInput();

                expect(mockOnSelectionRemove).not.toHaveBeenCalled();

                wrapper.find('input').simulate('keyDown', { key: 'Backspace' });

                expect(mockOnSelectionRemove).toHaveBeenCalled();
            });
            it('should not be called on deleting a suggestion by hitting backspace in an input field with value', () => {
                selectedSuggestions = SUGGESTIONS.slice(0, 5);
                wrapper.setProps({ selectedSuggestions });
                setFocusOnInput();
                wrapper.find('input').simulate('change', { target: { value: 'driver' } });

                expect(mockOnSelectionRemove).not.toHaveBeenCalled();

                wrapper.find('input').simulate('keyDown', { key: 'Backspace' });

                expect(mockOnSelectionRemove).not.toHaveBeenCalled();
            });
        });
        it('should call onInputValueChange when typing into input field', () => {
            expect(mockOnInputValueChange).not.toHaveBeenCalled();

            wrapper.find('input').simulate('change', { target: { value: 'driver' } });

            expect(mockOnInputValueChange).toHaveBeenCalled();
        });
        it('should clear the input field when a suggestion was selected', () => {
            const textInputValue = 'driver';
            suggestionsList = SUGGESTIONS.slice(1, 20);
            wrapper.setProps({ suggestions: suggestionsList });
            wrapper.find('input').simulate('change', { target: { value: textInputValue } });
            setFocusOnInput();

            expect(wrapper.find('input').props().value).toEqual(textInputValue);

            wrapper.find('li').first().children().simulate('click');

            expect(wrapper.find('input').props().value).toEqual('');
        });
    });
});
