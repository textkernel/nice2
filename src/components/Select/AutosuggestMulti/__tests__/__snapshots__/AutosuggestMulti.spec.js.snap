// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AutosuggestMulti rendering should initially render empty component correctly 1`] = `
<AutosuggestMulti
  inputPlaceholder="type here..."
  numberOfVisibleTags={3}
  onBlur={[MockFunction]}
  onInputValueChange={[MockFunction]}
  onSelectionChange={[MockFunction]}
  onSelectionRemove={[MockFunction]}
  selectedSuggestions={Array []}
  suggestionToString={[Function]}
  suggestions={Array []}
>
  <SelectBase
    blurredRenderer={[Function]}
    clearInputAfterSelection={true}
    clearTitle=""
    focusedRenderer={[Function]}
    inputRef={
      Object {
        "current": <input
          aria-autocomplete="list"
          aria-labelledby="downshift-0-label"
          autocomplete="off"
          class="AutosuggestMulti__input"
          data-lpignore="true"
          placeholder="type here..."
          value=""
        />,
      }
    }
    keepExpandedAfterSelection={true}
    listRenderer={[Function]}
    onBlur={[MockFunction]}
    onInputValueChange={[Function]}
    onSelectionChange={[MockFunction]}
    selectOnBlur={false}
    showClearButton={false}
    suggestionToString={[Function]}
    suggestions={Array []}
  >
    <div
      className="SelectBase"
    >
      <Downshift
        defaultHighlightedIndex={0}
        defaultIsOpen={false}
        environment={[Window]}
        getA11yStatusMessage={[Function]}
        inputValue=""
        itemToString={[Function]}
        onChange={[Function]}
        onInputValueChange={[Function]}
        onOuterClick={[Function]}
        onSelect={[Function]}
        onStateChange={[Function]}
        onUserAction={[Function]}
        scrollIntoView={[Function]}
        selectedItemChanged={[Function]}
        stateReducer={[Function]}
        suppressRefError={false}
      >
        <div
          aria-expanded={false}
          aria-haspopup="listbox"
          aria-labelledby="downshift-0-label"
          aria-owns={null}
          className="SelectBase__main"
          role="combobox"
        >
          <FieldWrapper
            className="SelectBase__field"
            clearLabel=""
            isFocused={false}
            onClear={[Function]}
            onClick={[Function]}
            showClearButton={false}
          >
            <div
              className="FieldWrapper SelectBase__field"
              onClick={[Function]}
            >
              <div
                className="FieldWrapper__content"
              >
                <div
                  className="AutosuggestMulti__wrapper"
                >
                  <input
                    aria-activedescendant={null}
                    aria-autocomplete="list"
                    aria-controls={null}
                    aria-labelledby="downshift-0-label"
                    autoComplete="off"
                    className="AutosuggestMulti__input"
                    data-lpignore={true}
                    onBlur={[Function]}
                    onChange={[Function]}
                    onFocus={[Function]}
                    onKeyDown={[Function]}
                    placeholder="type here..."
                    value=""
                  />
                </div>
                <List
                  aria-labelledby="downshift-0-label"
                  doSelectOnNavigate={false}
                  id="downshift-0-menu"
                  isControlledNavigation={true}
                  isDivided={false}
                  role="listbox"
                >
                  <ul
                    aria-labelledby="downshift-0-label"
                    className="List"
                    id="downshift-0-menu"
                    role="listbox"
                  />
                </List>
              </div>
            </div>
          </FieldWrapper>
        </div>
      </Downshift>
    </div>
  </SelectBase>
</AutosuggestMulti>
`;
